### DISCUSSION TITLE
Assigning orientations from file

### DISCUSSION THREAD

**Discussion by xshang93** [neperfepx/neper#325] [Q&A] [answered]
Hi, I'm trying to assign orientation for my 3-scale tessellation structure. Now I'm testing the `-ori 'file(filename)' ` option on a 2-scales tessellation.

First of all, is there a format that the file needs to follow or do I just do something like:
1 2 3
1 2 3
...
I used this format, output the .ori file, no error was generated. But the orientations are just not what i specified, they are still the same as the randomly generated orientation. 

Second, in the neper documentation on page 19, it says "if Rodrigues vector (rodrigues), the descriptor can be omitted." but on page 79, I found "When loading orientations from an external file, ‘file(file_name)’ assumes orientations described by Euler angles in Bunge convention (and active convention).". So which one of these is used by default?

Third, for a 2 or 3 scale tessellation, is there a way to find the grain IDs so I know which one to assign? Or do I just use 3 different files separated by :: ?

Many thanks in advance.
XS

**Comment by smallrain1996** [comment]
I've also noticed that for the same orifile, the following two commands get different .ori files
neper -T -n 2::4::3 \
-ori "file(orifile,des=rodrigues)" \
-format "tess,ori" \
-o test1

neper -T -n 24 \
-ori "file(orifile,des=rodrigues)" \
-format "tess,ori" \
-o test2

**Reply by xshang93** [reply]
I think that's expected. my guess is, when you give `-ori "file(orifile,des=rodrigues)"` in your three scale tessellation, it only reads it for the first level, and uses the default (which is random) for the second and third levels. At the end of the day, all the output orientations should only be defined by your third level -ori file, and should have nothing to do with the first and second level, yet you still need to specify them.

I still need to do more test on this.

XS

**Reply by rquey** [reply]
@xshang93 is correct. See https://github.com/rquey/neper/discussions/325#discussioncomment-1643961 for a proper solution.

**Comment by xshang93** [comment]
I did some more tests and I think I can answer some of my questions:

1. the file format is correct.
2. It's using rodrigues by default. I tested that out by doing a simple one scale tessellation.
3. I think the correct way to use `-ori` is, using a 2-scale tessellation as example, `-ori "file1::file2"`. But when I did this, neper stucked, and is using 99.9% of my CPU but not moving forward. The output from neper is attached below (I'm assigning random orientations for the second tessellation level):

```
========================    N   e   p   e   r    =======================
Info   : A software package for polycrystal generation and meshing.
Info   : Version 4.2.1-7
Info   : Built with: gsl|muparser|opengjk|openmp|nlopt|libscotch (full)
Info   : Running on 8 threads.
Info   : <https://neper.info>
Info   : Copyright (C) 2003-2021, and GNU GPL'd, by Romain Quey.
Info   : No initialization file found (/Users/Xiao/.neperrc').
Info   : ---------------------------------------------------------------
Info   : MODULE  -T loaded with arguments:
Info   : [ini file] (none)
Info   : [com line] -n 2::from_morpho -id 1::1 -morpho
         gg::lamellar(w=file(lamwidth),v=(1,0,0)) -ori
         file(ori,des=euler-bunge)::file(ori2,des=euler-bunge) -format
         tess,ori -o Polycrystal_laminated
Info   : ---------------------------------------------------------------
Info   : Reading input data...
Info   : Creating domain...
Info   : Creating tessellations...
Info   :   [level] Running in multiscale mode (2 scales)...
Info   :   [    1]  1/1 Info   :   - Setting seeds... 100%              
Info   :   - Generating crystal orientations...
Info   :     [i] Parsing file 'ori'...
Info   :     [i] Parsed file 'ori'.
Info   :   - Running tessellation...
Info   :     > Initial solution: f   =13.325428096
Info   :     > Iteration    218: fmin=1.800334242 f=1.800334242  
Info   :     > Final solution  : f   =1.800334242 (218 iterations)
Info   :     > Reached 'eps' criterion.
Info   :   [    2]  1/2 Info   :   - Running tessellation...            
Info   :   - Generating crystal orientations...
```

Can someone help with this?

**Reply by smallrain1996** [reply]
It may take a developer of the software to fundamentally reply to this technique, but I can offer two compromises.

If you are in need of subsequent calculations using Fepx, you can actually use a separate orientation file to provide orientation information regardless of how Neper generates the orientation.

If you just need to present the correct orientation information in the .tess file, you can write a script that modifies the content of the specified area in the .tess, because .tess is essentially a text file, and this is not difficult.

**Reply by xshang93** [reply]
These are very good points. Thanks. I think this is what I'll do for now.

XS

**Reply by rquey** [reply]
> I did some more tests and I think I can answer some of my questions:
> 
>     1. the file format is correct.
Yes.
> 
>     2. It's using rodrigues by default. I tested that out by doing a simple one scale tessellation.
Yes.
> 
>     3. I think the correct way to use `-ori` is, using a 2-scale tessellation as example, `-ori "file1::file2"`. But when I did this, neper stucked, and is using 99.9% of my CPU but not moving forward. The output from neper is attached below (I'm assigning random orientations for the second tessellation level):
> 
> [...]

Thanks for the report. Neper now generates an error when the input file is not properly formatted (instead of looping infinitely). Otherwise, see https://github.com/rquey/neper/discussions/325#discussioncomment-1643961 for a proper solution.

**Reply by xshang93** [reply]
Thank you. I had the feeling that not moving forward while eating up CPU signifies an error. Thank for confirming this. I guess the most recent issue I had, i.e. #326 , is due to such an issue too. However, I was not able to figure out what cost the error and it would be great if you could give some insight.

**Comment by rquey** [comment, marked as answer]
> Hi, I'm trying to assign orientation for my 3-scale tessellation structure. Now I'm testing the `-ori 'file(filename)' ` option on a 2-scales tessellation.
> 
> First of all, is there a format that the file needs to follow or do I just do something like: 1 2 3 1 2 3 ... I used this format, output the .ori file, no error was generated. But the orientations are just not what i specified, they are still the same as the randomly generated orientation.

The file should be formatted as a `.ori file`, so, just a list of orientations.
 
> Second, in the neper documentation on page 19, it says "if Rodrigues vector (rodrigues), the descriptor can be omitted." but on page 79, I found "When loading orientations from an external file, ‘file(file_name)’ assumes orientations described by Euler angles in Bunge convention (and active convention).". So which one of these is used by default?

Thanks for pointing this out. It is Rodrigues vectors. Fixed in version 4.2.1-14.

> Third, for a 2 or 3 scale tessellation, is there a way to find the grain IDs so I know which one to assign? Or do I just use 3 different files separated by :: ?
> 
> Many thanks in advance. XS

This may seem a bit tricky at first, but the way to do it is as follows. For a 2-scale tessellation for which you would want to specify the orientations of the cells at scale 2:
```
neper -T -n 3::4 -ori "random::file(scale2_file)"
```
where `scale2_file` is a multiscale file that should contain, for each scale-1 cell, the orientation distribution. It could be `random` (as at scale 1), but it can also be of the type `file(...)` to load custom orientations, or any other input described in option `-ori`. In your case, the content of `scale2_file` could be:
```
1 file(cell1)
2 file(cell2)
3 file(cell3)
```
where `cell1` would be an orientation file, which would contain the orientations of the 4 cells subdividing the 1st scale-1 cell, e.g.,
```
   0.704910491805    1.813296457703    0.720973368800
  -1.755963753575    1.732321781503    4.425698589080
   2.557193098119   -1.540193224736   -0.227983822696
  -0.349045017160    0.088659451979   -0.322358107095
```
Same for `cell2` and `cell3`.

This can be generalized to 3 scales and more.

**Reply by xshang93** [reply]
Thanks for the reply. I'll test it out tonight!

**Reply by janzenchoi** [reply]
Hello @rquey,

I just tried the above commands with the newest version of Neper. However, I got a segmentation fault from running them.

![image](https://user-images.githubusercontent.com/75835751/196078434-96d674a6-223b-4a54-b649-295a11afbb3d.png)

Are those commands / formats no longer supported in the newest version of Neper? Is there a new set of commands I could use instead?

Thanks.

**Reply by rquey** [reply]
Please provide the input  necessary to reproduce the bug (`index.txt`) and copy text as text (using Markdown formatting), not images... This is cleaner and saves us from re-typing commands...

**Reply by janzenchoi** [reply]
Hello @rquey,

Sure thing. The commands and files should be quite similar to the ones you suggested above.

The `index.txt` file contains:

```
1 file(cell1.txt)
2 file(cell2.txt)
3 file(cell3.txt)
```

The `cell1.txt`, `cell2.txt`, and `cell3.txt` files all contain:
```
   0.704910491805    1.813296457703    0.720973368800
  -1.755963753575    1.732321781503    4.425698589080
   2.557193098119   -1.540193224736   -0.227983822696
  -0.349045017160    0.088659451979   -0.322358107095
```

Finally, after creating those files, I ran the following command.

`neper -T -n 3::4 -ori "random::file(index.txt)" -o img`

I seem to get a segmentation fault from it.

I was trying to apply crystal orientations to a multiscale and got a similar error, which is how I found this discussion post.

**Comment by rquey** [comment]
Here is what I get:
```
========================    N   e   p   e   r    =======================
Info   : A software package for polycrystal generation and meshing.
Info   : Version 4.5.1-2
Info   : Built with: gsl|muparser|opengjk|openmp|nlopt|libscotch (full)
Info   : Running on 8 threads.
Info   : <https://neper.info>
Info   : Copyright (C) 2003-2022, and GNU GPL'd, by Romain Quey.
Info   : Loading initialization file `/home/rquey/.neperrc'...
Info   : ---------------------------------------------------------------
Info   : MODULE  -T loaded with arguments:
Info   : [ini file] -tesrformat ascii
Info   : [com line] -n 3::4 -ori random::file(index.txt) -o img
Info   : ---------------------------------------------------------------
Info   : Reading input data...
Info   : Creating domain...
Info   : Creating tessellations...
Info   :   [level] Running in multiscale mode (2 scales)...
Info   :   [    1]  1/1 Info   :   - Setting seeds...                   
Info   :   - Generating crystal orientations...
Info   :   - Running tessellation...
Info   :   [    2]  1/3 Info   :   - Setting seeds...                   
Info   :   - Generating crystal orientations...
Info   :     [i] Parsing file `cell1.txt'...
Info   :     [i] Parsed file `cell1.txt'.
Info   :   - Running tessellation...
Info   :   [    2]  2/3 Info   :   - Setting seeds...                   
Info   :   - Generating crystal orientations...
Info   :     [i] Parsing file `cell2.txt'...
Info   :     [i] Parsed file `cell2.txt'.
Info   :   - Running tessellation...
Info   :   [    2]  3/3 Info   :   - Setting seeds...                   
Info   :   - Generating crystal orientations...
Info   :     [i] Parsing file `cell3.txt'...
Info   :     [i] Parsed file `cell3.txt'.
Info   :   - Running tessellation...
Info   : Flattening 2-scale tessellation...
Info   :   [level] Flattening 2-scale tessellation...
Info   :   [    1]  1/1: id=1                                           
Info   : Writing results...
Info   :     [o] Writing file `img.tess'...
Info   :     [o] Wrote file `img.tess'.
Info   : Elapsed time: 0.051 secs.
========================================================================
```

**Reply by janzenchoi** [reply]
In text form, mine is:

```
========================    N   e   p   e   r    =======================
Info   : A software package for polycrystal generation and meshing.
Info   : Version 4.5.1-2
Info   : Built with: gsl|muparser|opengjk|openmp|nlopt|libscotch (full)
Info   : Running on 16 threads.
Info   : <https://neper.info>
Info   : Copyright (C) 2003-2022, and GNU GPL'd, by Romain Quey.
Info   : No initialization file found (`/home/janzen/.neperrc').
Info   : ---------------------------------------------------------------
Info   : MODULE  -T loaded with arguments:
Info   : [ini file] (none)
Info   : [com line] -n 3::4 -ori random::file(index.txt) -o img
Info   : ---------------------------------------------------------------
Info   : Reading input data...
Info   : Creating domain...
Info   : Creating tessellations...
Info   :   [level] Running in multiscale mode (2 scales)...
Info   :   [    1]  1/1 Info   :   - Setting seeds...
Info   :   - Generating crystal orientations...
Info   :   - Running tessellation...
Info   :   [    2]  1/3 Info   :   - Setting seeds...
Info   :   - Generating crystal orientations...
Info   :     [i] Parsing file `cell1.txt'...
Info   :     [i] Parsed file `cell1.txt'.
Info   :   - Running tessellation...
Info   :   [    2]  2/3 Info   :   - Setting seeds...
Info   :   - Generating crystal orientations...
Info   :     [i] Parsing file `cell2.txt'...
Info   :     [i] Parsed file `cell2.txt'.
Info   :   - Running tessellation...
Info   :   [    2]  3/3 Info   :   - Setting seeds...
Info   :   - Generating crystal orientations...
Segmentation fault
```

That really is weird. I think the only difference in our outputs before the segmentation fault is the `ini file`.

**Reply by rquey** [reply]
Please attach your txt files.

**Reply by janzenchoi** [reply]
Sure thing. 
[files.zip](https://github.com/neperfepx/neper/files/9828832/files.zip)
